pkg/virt-handler/vm.go

```diff
@@ -2117,6 +2082,51 @@ func (d *VirtualMachineController) processVmShutdown(vmi *v1.VirtualMachineInsta
 	return nil
 }
 
+func (d *VirtualMachineController) handleVMIShutdown(vmi *v1.VirtualMachineInstance, domain *api.Domain, client cmdclient.LauncherClient, timeLeft int64) error {
+	if domain.Status.Status != api.Shutdown {
+		return d.shutdownVMI(vmi, client, timeLeft)
+	}
+	log.Log.V(4).Object(vmi).Infof("%s is already shutting down.", vmi.GetObjectMeta().GetName())
+	return nil
+}
+
+func (d *VirtualMachineController) shutdownVMI(vmi *v1.VirtualMachineInstance, client cmdclient.LauncherClient, timeLeft int64) error {
+	err := client.ShutdownVirtualMachine(vmi)
+	if !isCausedByDisconnect(err) {
+		return err
+	}
+
+	log.Log.Object(vmi).Infof("Signaled graceful shutdown for %s", vmi.GetObjectMeta().GetName())
+
+	// Make sure that we don't hot-loop in case we send the first domain notification
+	if timeLeft == -1 {
+		timeLeft = 5
+		if vmi.Spec.TerminationGracePeriodSeconds != nil && *vmi.Spec.TerminationGracePeriodSeconds < timeLeft {
+			timeLeft = *vmi.Spec.TerminationGracePeriodSeconds
+		}
+	}
+	// In case we have a long grace period, we want to resend the graceful shutdown every 5 seconds
+	// That's important since a booting OS can miss ACPI signals
+	if timeLeft > 5 {
+		timeLeft = 5
+	}
+
+	// pending graceful shutdown.
+	d.Queue.AddAfter(controller.VirtualMachineInstanceKey(vmi), time.Duration(timeLeft)*time.Second)
+	d.recorder.Event(vmi, k8sv1.EventTypeNormal, v1.ShuttingDown.String(), VMIGracefulShutdown)
+	return nil
+}
+
+func isCausedByDisconnect(err error) bool {
+	// Only report err if it wasn't the result of a disconnect.
+	//
+	// Both virt-launcher and virt-handler are trying to destroy
+	// the VirtualMachineInstance at the same time. It's possible the client may get
+	// disconnected during the kill request, which shouldn't be
+	// considered an error.
+	return err == nil || cmdclient.IsDisconnected(err)
```

Comment:
I find it a bit confusing that in case there is no error, `isCausedByDisconnect` returns true.

Reply:
I agree that it is confusing. 
Changed name of the function. 
Personally, I don't like that either, but nothing more suitable comes to my mind.

Reply:
Finally, I inlined that condition, as Radim suggested.